File: flux/apps/games/futboss-ai/packages/shared/src/types/match.ts
// FutBoss AI - Match Types
// Author: Bruno Lucena (bruno@lucena.cloud)

export type MatchStatus = 'pending' | 'in_progress' | 'finished' | 'cancelled';
export type EventType = 'goal' | 'assist' | 'yellow_card' | 'red_card' | 'substitution' | 'injury' | 'penalty' | 'save' | 'foul';

export interface MatchEvent {
  minute: number;
  eventType: EventType;
  playerId: string;
  teamId: string;
  description: string;
  aiNarration?: string;
}

export interface MatchState {
  minute: number;
  homeScore: number;
  awayScore: number;
  possessionHome: number;
  possessionAway: number;
  shotsHome: number;
  shotsAway: number;
  isPaused: boolean;
}

export interface Match {
  id: string;
  homeTeamId: string;
  awayTeamId: string;
  homeScore: number;
  awayScore: number;
  status: MatchStatus;
  events: MatchEvent[];
  state: MatchState;
  startedAt?: Date;
  finishedAt?: Date;
  createdAt: Date;
}

export interface MatchCreate {
  homeTeamId: string;
  awayTeamId: string;
}

export function isHomeWinner(match: Match): boolean | null {
  if (match.status !== 'finished') return null;
  if (match.homeScore === match.awayScore) return null;
  return match.homeScore > match.awayScore;
}

export function isDraw(match: Match): boolean {
  return match.status === 'finished' && match.homeScore === match.awayScore;
}

export function totalGoals(match: Match): number {
  return match.homeScore + match.awayScore;
}

File: flux/apps/games/futboss-ai/apps/api/app/models/match.py
# FutBoss AI - Match Model
# Author: Bruno Lucena (bruno@lucena.cloud)

from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime
from enum import Enum


class MatchStatus(str, Enum):
    PENDING = "pending"
    IN_PROGRESS = "in_progress"
    FINISHED = "finished"
    CANCELLED = "cancelled"


class EventType(str, Enum):
    GOAL = "goal"
    ASSIST = "assist"
    YELLOW_CARD = "yellow_card"
    RED_CARD = "red_card"
    SUBSTITUTION = "substitution"
    INJURY = "injury"
    PENALTY = "penalty"
    SAVE = "save"
    FOUL = "foul"


class MatchEvent(BaseModel):
    minute: int = Field(ge=0, le=120)
    event_type: EventType
    player_id: str
    team_id: str
    description: str
    ai_narration: Optional[str] = None


class MatchState(BaseModel):
    """Current state of a match for real-time updates"""
    minute: int = 0
    home_score: int = 0
    away_score: int = 0
    possession_home: int = 50
    possession_away: int = 50
    shots_home: int = 0
    shots_away: int = 0
    is_paused: bool = False


class MatchCreate(BaseModel):
    home_team_id: str
    away_team_id: str


class Match(BaseModel):
    id: Optional[str] = Field(default=None, alias="_id")
    home_team_id: str
    away_team_id: str
    home_score: int = 0
    away_score: int = 0
    status: MatchStatus = MatchStatus.PENDING
    events: List[MatchEvent] = Field(default_factory=list)
    state: MatchState = Field(default_factory=MatchState)
    started_at: Optional[datetime] = None
    finished_at: Optional[datetime] = None
    created_at: datetime = Field(default_factory=datetime.utcnow)

    class Config:
        populate_by_name = True

    def is_home_winner(self) -> Optional[bool]:
        if self.status != MatchStatus.FINISHED:
            return None
        if self.home_score > self.away_score:
            return True
        elif self.home_score < self.away_score:
            return False
        return None  # Draw

